### ISSUE TITLE
Wrong type in distribution query unshard visitor

### ISSUE THREAD

**Issue by danielmewes** [rethinkdb/rethinkdb#4233]
Cannot reproduce, but got it once:

```
error: Error in src/rdb_protocol/protocol.cc at line 770:
error: Guarantee failed: [result != __null] Bad boost::get
error: Backtrace:
error: Thu May 14 16:45:04 2015
       
       1: format_backtrace(bool) at backtrace.cc:214
       2: report_fatal_error(char const*, int, char const*, ...) at errors.cc:83
       3: rdb_r_unshard_visitor_t::operator()(distribution_read_t const&) at protocol.cc:770
       4: read_t::unshard(read_response_t*, unsigned long, read_response_t*, rdb_context_t*, signal_t*) const at variant.hpp:832
       5: table_query_client_t::dispatch_outdated_read(read_t const&, read_response_t*, signal_t*) at stl_vector.h:350
       6: fetch_distribution(uuid_u const&, real_reql_cluster_interface_t*, signal_t*, std::map<store_key_t, long, std::less<store_key_t>, std::allocator<std::pair<store_key_t const, long> > >*) at split_points.cc:237
       7: real_reql_cluster_interface_t::table_estimate_doc_counts(counted_t<ql::db_t const>, name_string_t const&, ql::env_t*, std::vector<long, std::allocator<long> >*, std::string*) at stl_vector.h:571
       8: artificial_reql_cluster_interface_t::table_estimate_doc_counts(counted_t<ql::db_t const>, name_string_t const&, ql::env_t*, std::vector<long, std::allocator<long> >*, std::string*) at artificial_reql_cluster_interface.cc:169
       9: ql::info_term_t::val_info(ql::scope_env_t*, scoped_ptr_t<ql::val_t>) const at counted.hpp:60
       10: ql::info_term_t::eval_impl(ql::scope_env_t*, ql::args_t*, ql::eval_flags_t) const at datum.hpp:96
       11: ql::op_term_t::term_eval(ql::scope_env_t*, ql::eval_flags_t) const at scoped.hpp:70
       12: ql::runtime_term_t::eval(ql::scope_env_t*, ql::eval_flags_t) const at scoped.hpp:25
       13: ql::op_term_t::maybe_grouped_data(ql::scope_env_t*, ql::argvec_t*, ql::eval_flags_t, counted_t<ql::grouped_data_t>*, scoped_ptr_t<ql::val_t>*) const at counted.hpp:60
       14: ql::op_term_t::term_eval(ql::scope_env_t*, ql::eval_flags_t) const at op.cc:199
       15: ql::runtime_term_t::eval(ql::scope_env_t*, ql::eval_flags_t) const at scoped.hpp:25
       16: ql::op_term_t::maybe_grouped_data(ql::scope_env_t*, ql::argvec_t*, ql::eval_flags_t, counted_t<ql::grouped_data_t>*, scoped_ptr_t<ql::val_t>*) const at counted.hpp:60
       17: ql::op_term_t::term_eval(ql::scope_env_t*, ql::eval_flags_t) const at op.cc:199
       18: ql::runtime_term_t::eval(ql::scope_env_t*, ql::eval_flags_t) const at scoped.hpp:25
       19: ql::op_term_t::maybe_grouped_data(ql::scope_env_t*, ql::argvec_t*, ql::eval_flags_t, counted_t<ql::grouped_data_t>*, scoped_ptr_t<ql::val_t>*) const at counted.hpp:60
       20: ql::op_term_t::term_eval(ql::scope_env_t*, ql::eval_flags_t) const at op.cc:199
       21: ql::runtime_term_t::eval(ql::scope_env_t*, ql::eval_flags_t) const at scoped.hpp:25
       22: ql::make_obj_term_t::term_eval(ql::scope_env_t*, ql::eval_flags_t) const at datum_terms.cc:87
       23: ql::runtime_term_t::eval(ql::scope_env_t*, ql::eval_flags_t) const at scoped.hpp:25
       24: ql::args_t::arg(ql::scope_env_t*, unsigned long, ql::eval_flags_t) at counted.hpp:60
       25: ql::merge_term_t::obj_eval(ql::scope_env_t*, ql::args_t*, scoped_ptr_t<ql::val_t> const&) const at obj_or_seq.cc:224
       26: ./rethinkdb() [0xadb535] at 0xadb535 ()
       27: ql::obj_or_seq_op_impl_t::eval_impl_dereferenced(ql::term_t const*, ql::scope_env_t*, ql::args_t*, scoped_ptr_t<ql::val_t> const&, std::function<scoped_ptr_t<ql::val_t> ()>) const at datum.hpp:96
       28: ql::obj_or_seq_op_term_t::eval_impl(ql::scope_env_t*, ql::args_t*, ql::eval_flags_t) const at functional:1736
       29: ql::op_term_t::term_eval(ql::scope_env_t*, ql::eval_flags_t) const at scoped.hpp:70
       30: ql::runtime_term_t::eval(ql::scope_env_t*, ql::eval_flags_t) const at scoped.hpp:25
       31: ql::reql_func_t::call(ql::env_t*, std::vector<ql::datum_t, std::allocator<ql::datum_t> > const&, ql::eval_flags_t) const at datum.hpp:96
       32: ql::funcall_term_t::eval_impl(ql::scope_env_t*, ql::args_t*, ql::eval_flags_t) const at control.cc:122
       33: ql::op_term_t::term_eval(ql::scope_env_t*, ql::eval_flags_t) const at scoped.hpp:70
       34: ql::runtime_term_t::eval(ql::scope_env_t*, ql::eval_flags_t) const at scoped.hpp:25
       35: ql::query_cache_t::ref_t::run(ql::env_t*, Response*) at query_cache.cc:278
       36: ql::query_cache_t::ref_t::fill_response(Response*) at query_cache.cc:236
       37: ql::run(ql::query_id_t&&, ql::protob_t<Query>, Response*, ql::query_cache_t*, signal_t*) at term.cc:253
       38: rdb_query_server_t::run_query(ql::query_id_t&&, ql::protob_t<Query> const&, Response*, ql::query_cache_t*, signal_t*) at counted.hpp:60
       39: query_server_t::handle(http_req_t const&, http_res_t*, signal_t*) at protob.cc:706
       40: routing_http_app_t::handle(http_req_t const&, http_res_t*, signal_t*) at basic_string.h:288
       41: routing_http_app_t::handle(http_req_t const&, http_res_t*, signal_t*) at basic_string.h:288
       42: http_server_t::handle_conn(scoped_ptr_t<linux_tcp_conn_descriptor_t> const&, auto_drainer_t::lock_t) at basic_string.h:542
       43: std::_Function_handler<void (scoped_ptr_t<linux_tcp_conn_descriptor_t>&), boost::_bi::bind_t<void, boost::_mfi::mf2<void, http_server_t, scoped_ptr_t<linux_tcp_conn_descriptor_t> const&, auto_drainer_t::lock_t>, boost::_bi::list3<boost::_bi::value<http_server_t*>, boost::arg<1>, boost::_bi::value<auto_drainer_t::lock_t> > > >::_M_invoke(std::_Any_data const&, scoped_ptr_t<linux_tcp_conn_descriptor_t>&) at mem_fn_template.hpp:280
       44: linux_nonthrowing_tcp_listener_t::handle(int) at scoped.hpp:70
       45: coro_t::run() at coroutines.cc:214
error: Exiting.
Trace/breakpoint trap (core dumped)
```

@timmaxw you mentioned you had also gotten this at some point, right?

**Comment by timmaxw** [comment]
Yeah, I've seen this multiple times before in the shard fuzzer. I have no idea what causes it.

**Comment by timmaxw** [comment]
Fixed in ad61fc3e04506c948d490bf881d17a34fd4fdbe5.
